src/rules.jl

```diff
@@ -90,6 +90,7 @@ else
 end
 @define_diffrule Base.hypot(x, y)  = :( $x / hypot($x, $y)                                      ), :(  $y / hypot($x, $y)                                                     )
 @define_diffrule Base.log(b, x)    = :( log($x) * inv(-log($b)^2 * $b)                          ), :( inv($x) / log($b)                                                       )
+@define_diffrule Base.ldexp(x, y)  = :( 2^$y                                                    ), :(log(2) * $x * 2^$y)
```

Comment:
Intuitively, I would say that the derivative of `ldexp` with respect to the second argument is not defined - in ChainRules it would be `NoTangent` (and actually it is: https://github.com/JuliaDiff/ChainRules.jl/blob/ffbaa5fecca8da39f20aeb66cc6e4edf3e0c3f11/src/rulesets/Base/base.jl#L105) and AFAIK in DiffRules this is indicated by `NaN`.

So in my opinion the rule should be defined as
```suggestion
@define_diffrule Base.ldexp(x, y)  = :( exp2($y) ), :NaN
```
Here `exp2` avoids overflow:
```julia
julia> ldexp(1.0, 64)
1.8446744073709552e19

julia> 2^64
0

julia> exp2(64)
1.8446744073709552e19
```

Reply:
Dear @devmotion 
thank you for the comments. The derivative of 2^x is defined: note that 2^x = exp(x\*log(2)), deriving exp(a\*x) w.r.t. x is just a\*exp(a\*x)

Thanks for pointing out the `exp2` feature of avoiding overflows!

Also, I'll note that I am still pretty new and unfamiliar with the workflow in git so please advice if I am missing anything I am supposed to do. Thanks!

Reply:
Sure, one can differentiate `x -> 2^x`. But the crucial point here is that `ldexp` only accepts integers as second argument, and in automatic differentiation, as shown also by the ChainRules definition linked above, integers are considered to be a discrete set and hence non-differentiable. Of course, they could be embedded in the real line but perturbing the input would lead to values that the function can't be called with: `ldexp` only accepts `Integer`s as second argument and errors if it is called with non-integer values (BTW this is also the reason for the test errors.)

This is also explained in the ChainRules documentation: https://juliadiff.org/ChainRulesCore.jl/stable/FAQ.html#faq_abstract_zero

> Also, I'll note that I am still pretty new and unfamiliar with the workflow in git so please advice if I am missing anything I am supposed to do.

In the web interface you can accept my suggestion above by clicking on the "commit suggestion" button below it (if I managed to convince you that it is an improvement :wink:). Otherwise, the main thing left to do is to fix the test errors. This can be done by updating the tests in `test/runtests.jl` and ensuring that the second argument is always an integer, similar to how `rem2pi` is handled: https://github.com/JuliaDiff/DiffRules.jl/blob/68cd4f2416c68bb9ed349bf92e2acc9221591c49/test/runtests.jl#L79